File: ingestion/storage.py
"""
Storage layer.

Choice: PostgreSQL over MongoDB.

Justification -- this pipeline's data is NOT uniformly unstructured, and
that's the deciding factor:
  - Financial statements and key stats are genuinely tabular/structured
    (fixed line items, numeric values) -- a natural fit for relational
    columns, and Role 2's ratio agent will want to query/join/aggregate
    these numerically (e.g. "average current ratio across peers"), which
    Postgres does far more naturally than Mongo's aggregation pipeline.
  - Filing sections, transcript utterances, and news articles are
    semi-structured text blobs, but Postgres's JSONB columns handle that
    fine while still letting the CitationMetadata fields (ticker, url,
    section_id, retrieved_at) live as real indexed columns -- so "find
    every section from AAPL's most recent 10-K" is a normal indexed query,
    not a full collection scan.
  - Postgres full-text search (tsvector) covers the "search across filing
    text / news snippets" need without introducing a second datastore.
  - A single relational schema is easier for Role 2 and Role 3 to reason
    about when reconstructing a citation across statement + filing +
    transcript + news in one query, which is exactly the join-heavy
    access pattern this pipeline needs.

MongoDB would be the better call if the data were mostly free-form
documents with no numeric analysis burden -- it isn't, here.
"""

from __future__ import annotations

import json
from datetime import datetime
from typing import Any

from .schemas import FilingDocument, FinancialStatement, NewsArticle, TranscriptDocument

try:
    import psycopg2
    import psycopg2.extras
except ImportError:
    psycopg2 = None


DDL = """
CREATE TABLE IF NOT EXISTS financial_statements (
    id SERIAL PRIMARY KEY,
    ticker TEXT NOT NULL,
    statement_type TEXT NOT NULL,
    fiscal_period TEXT NOT NULL,
    source_name TEXT NOT NULL,
    source_url TEXT,
    retrieved_at TIMESTAMPTZ NOT NULL,
    line_items JSONB NOT NULL,
    missing_fields JSONB NOT NULL,
    UNIQUE (ticker, statement_type, fiscal_period)
);
CREATE INDEX IF NOT EXISTS idx_fin_stmt_ticker ON financial_statements (ticker);

CREATE TABLE IF NOT EXISTS filing_sections (
    id SERIAL PRIMARY KEY,
    ticker TEXT NOT NULL,
    filing_type TEXT NOT NULL,
    accession_number TEXT NOT NULL,
    filed_date TIMESTAMPTZ NOT NULL,
    section_name TEXT NOT NULL,
    section_id TEXT NOT NULL,
    paragraph_index INT NOT NULL,
    source_url TEXT NOT NULL,
    retrieved_at TIMESTAMPTZ NOT NULL,
    text TEXT NOT NULL,
    text_search tsvector GENERATED ALWAYS AS (to_tsvector('english', text)) STORED,
    UNIQUE (ticker, accession_number, section_id)
);
CREATE INDEX IF NOT EXISTS idx_filing_ticker ON filing_sections (ticker, filing_type);
CREATE INDEX IF NOT EXISTS idx_filing_text_search ON filing_sections USING GIN (text_search);

CREATE TABLE IF NOT EXISTS transcript_utterances (
    id SERIAL PRIMARY KEY,
    ticker TEXT NOT NULL,
    fiscal_quarter TEXT,
    call_date TIMESTAMPTZ,
    available BOOLEAN NOT NULL,
    unavailable_reason TEXT,
    speaker TEXT,
    section TEXT,
    sequence INT,
    section_id TEXT,
    source_url TEXT,
    retrieved_at TIMESTAMPTZ,
    text TEXT
);
CREATE INDEX IF NOT EXISTS idx_transcript_ticker ON transcript_utterances (ticker, fiscal_quarter);

CREATE TABLE IF NOT EXISTS news_articles (
    id SERIAL PRIMARY KEY,
    ticker TEXT NOT NULL,
    title TEXT NOT NULL,
    source_publication TEXT NOT NULL,
    published_at TIMESTAMPTZ NOT NULL,
    url TEXT NOT NULL,
    snippet TEXT,
    dedupe_key TEXT NOT NULL,
    retrieved_at TIMESTAMPTZ NOT NULL,
    UNIQUE (dedupe_key)
);
CREATE INDEX IF NOT EXISTS idx_news_ticker ON news_articles (ticker, published_at);
"""


class Storage:
    def __init__(self, dsn: str):
        if psycopg2 is None:
            raise RuntimeError("psycopg2 is not installed in this environment")
        self.conn = psycopg2.connect(dsn)
        self.conn.autocommit = True

    def init_schema(self) -> None:
        with self.conn.cursor() as cur:
            cur.execute(DDL)

    def write_financial_statement(self, stmt: FinancialStatement) -> None:
        with self.conn.cursor() as cur:
            cur.execute(
                """
                INSERT INTO financial_statements
                    (ticker, statement_type, fiscal_period, source_name, source_url,
                     retrieved_at, line_items, missing_fields)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
                ON CONFLICT (ticker, statement_type, fiscal_period) DO UPDATE SET
                    line_items = EXCLUDED.line_items,
                    missing_fields = EXCLUDED.missing_fields,
                    retrieved_at = EXCLUDED.retrieved_at
                """,
                (
                    stmt.metadata.ticker,
                    stmt.statement_type.value,
                    stmt.fiscal_period,
                    stmt.metadata.source_name,
                    stmt.metadata.source_url,
                    stmt.metadata.retrieved_at,
                    json.dumps([li.model_dump(mode="json") for li in stmt.line_items]),
                    json.dumps(stmt.missing_fields),
                ),
            )

    def write_filing(self, filing: FilingDocument) -> None:
        with self.conn.cursor() as cur:
            rows = [
                (
                    filing.ticker,
                    filing.filing_type,
                    filing.accession_number,
                    filing.filed_date,
                    section.section_name,
                    section.metadata.section_id,
                    section.paragraph_index,
                    section.metadata.source_url,
                    section.metadata.retrieved_at,
                    section.text,
                )
                for section in filing.sections
            ]
            psycopg2.extras.execute_values(
                cur,
                """
                INSERT INTO filing_sections
                    (ticker, filing_type, accession_number, filed_date, section_name,
                     section_id, paragraph_index, source_url, retrieved_at, text)
                VALUES %s
                ON CONFLICT (ticker, accession_number, section_id) DO NOTHING
                """,
                rows,
            )

    def write_transcript(self, transcript: TranscriptDocument) -> None:
        with self.conn.cursor() as cur:
            if not transcript.available:
                cur.execute(
                    """
                    INSERT INTO transcript_utterances
                        (ticker, fiscal_quarter, call_date, available, unavailable_reason)
                    VALUES (%s, %s, %s, %s, %s)
                    """,
                    (transcript.ticker, transcript.fiscal_quarter, transcript.call_date,
                     False, transcript.unavailable_reason),
                )
                return
            rows = [
                (
                    transcript.ticker, transcript.fiscal_quarter, transcript.call_date, True, None,
                    u.speaker, u.section.value, u.sequence, u.metadata.section_id,
                    u.metadata.source_url, u.metadata.retrieved_at, u.text,
                )
                for u in transcript.utterances
            ]
            psycopg2.extras.execute_values(
                cur,
                """
                INSERT INTO transcript_utterances
                    (ticker, fiscal_quarter, call_date, available, unavailable_reason,
                     speaker, section, sequence, section_id, source_url, retrieved_at, text)
                VALUES %s
                """,
                rows,
            )

    def write_news(self, articles: list[NewsArticle]) -> None:
        if not articles:
            return
        with self.conn.cursor() as cur:
            rows = [
                (
                    a.metadata.ticker, a.title, a.source_publication, a.published_at,
                    a.url, a.snippet, a.dedupe_key, a.metadata.retrieved_at,
                )
                for a in articles
            ]
            psycopg2.extras.execute_values(
                cur,
                """
                INSERT INTO news_articles
                    (ticker, title, source_publication, published_at, url, snippet,
                     dedupe_key, retrieved_at)
                VALUES %s
                ON CONFLICT (dedupe_key) DO NOTHING
                """,
                rows,
            )

    def reconstruct_citation(self, table: str, row_id: int) -> dict[str, Any] | None:
        """Sanity-check helper: given any row, return enough fields to build
        a full citation (source, section, date, URL). Used in tests to prove
        nothing was dropped between agent output and storage."""
        allowed = {"financial_statements", "filing_sections", "transcript_utterances", "news_articles"}
        if table not in allowed:
            raise ValueError(f"unknown table {table}")
        with self.conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
            cur.execute(f"SELECT * FROM {table} WHERE id = %s", (row_id,))  # noqa: S608 -- table is allow-listed above
            return cur.fetchone()

File: cpp_pipeline/src/storage.cpp
#include "storage.h"
#include <sqlite3.h>
#include <iostream>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

Storage::Storage(const std::string& db_path) {
    if (sqlite3_open(db_path.c_str(), &db_) != SQLITE_OK) {
        std::cerr << "[ERROR] Cannot open SQLite DB: " << sqlite3_errmsg(db_) << std::endl;
        db_ = nullptr;
    }
}

Storage::~Storage() {
    if (db_) {
        sqlite3_close(db_);
    }
}

void Storage::execute(const std::string& sql) {
    if (!db_) return;
    char* err_msg = nullptr;
    if (sqlite3_exec(db_, sql.c_str(), nullptr, nullptr, &err_msg) != SQLITE_OK) {
        std::cerr << "[ERROR] SQLite error: " << err_msg << "\nSQL: " << sql << std::endl;
        sqlite3_free(err_msg);
    }
}

void Storage::init_schema() {
    execute(R"(
        CREATE TABLE IF NOT EXISTS financial_statements (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            ticker TEXT NOT NULL,
            statement_type TEXT NOT NULL,
            fiscal_period TEXT NOT NULL,
            line_items TEXT NOT NULL,
            UNIQUE (ticker, statement_type, fiscal_period)
        );
        CREATE TABLE IF NOT EXISTS filing_sections (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            ticker TEXT NOT NULL,
            filing_type TEXT NOT NULL,
            filed_date TEXT NOT NULL,
            section_name TEXT NOT NULL,
            text TEXT NOT NULL
        );
        CREATE TABLE IF NOT EXISTS transcript_utterances (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            ticker TEXT NOT NULL,
            speaker TEXT NOT NULL,
            section TEXT NOT NULL,
            text TEXT NOT NULL
        );
        CREATE TABLE IF NOT EXISTS news_articles (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            ticker TEXT NOT NULL,
            title TEXT NOT NULL,
            source_publication TEXT NOT NULL,
            published_at TEXT NOT NULL,
            url TEXT NOT NULL,
            snippet TEXT,
            UNIQUE (url)
        );
    )");
}

void Storage::write_financial_statement(const std::string& ticker, const FinancialStatement& stmt) {
    if (!db_) return;
    std::string sql = "INSERT OR REPLACE INTO financial_statements (ticker, statement_type, fiscal_period, line_items) VALUES (?, ?, ?, ?);";
    sqlite3_stmt* res;
    if (sqlite3_prepare_v2(db_, sql.c_str(), -1, &res, 0) == SQLITE_OK) {
        sqlite3_bind_text(res, 1, ticker.c_str(), -1, SQLITE_STATIC);
        sqlite3_bind_text(res, 2, stmt.statement_type.c_str(), -1, SQLITE_STATIC);
        sqlite3_bind_text(res, 3, stmt.fiscal_period.c_str(), -1, SQLITE_STATIC);
        
        json j = json::array();
        for (const auto& li : stmt.line_items) {
            j.push_back({{"name", li.name}, {"value", li.value.value_or(0.0)}, {"unit", li.unit}});
        }
        std::string j_str = j.dump();
        sqlite3_bind_text(res, 4, j_str.c_str(), -1, SQLITE_TRANSIENT);
        
        sqlite3_step(res);
        sqlite3_finalize(res);
    }
}

void Storage::write_filing(const std::string& ticker, const EdgarDocument& doc) {
    if (!db_) return;
    std::string sql = "INSERT INTO filing_sections (ticker, filing_type, filed_date, section_name, text) VALUES (?, ?, ?, ?, ?);";
    sqlite3_stmt* res;
    if (sqlite3_prepare_v2(db_, sql.c_str(), -1, &res, 0) == SQLITE_OK) {
        for (const auto& sec : doc.sections) {
            sqlite3_bind_text(res, 1, ticker.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 2, "quarterly_results", -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 3, doc.filed_date.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 4, sec.name.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 5, sec.text.c_str(), -1, SQLITE_STATIC);
            sqlite3_step(res);
            sqlite3_reset(res);
        }
        sqlite3_finalize(res);
    }
}

void Storage::write_transcript(const std::string& ticker, const TranscriptDocument& doc) {
    if (!db_) return;
    if (!doc.available) return;
    std::string sql = "INSERT INTO transcript_utterances (ticker, speaker, section, text) VALUES (?, ?, ?, ?);";
    sqlite3_stmt* res;
    if (sqlite3_prepare_v2(db_, sql.c_str(), -1, &res, 0) == SQLITE_OK) {
        for (const auto& u : doc.utterances) {
            sqlite3_bind_text(res, 1, ticker.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 2, u.speaker.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 3, u.section.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 4, u.text.c_str(), -1, SQLITE_STATIC);
            sqlite3_step(res);
            sqlite3_reset(res);
        }
        sqlite3_finalize(res);
    }
}

void Storage::write_news(const std::string& ticker, const std::vector<NewsArticle>& articles) {
    if (!db_) return;
    std::string sql = "INSERT OR IGNORE INTO news_articles (ticker, title, source_publication, published_at, url, snippet) VALUES (?, ?, ?, ?, ?, ?);";
    sqlite3_stmt* res;
    if (sqlite3_prepare_v2(db_, sql.c_str(), -1, &res, 0) == SQLITE_OK) {
        for (const auto& a : articles) {
            sqlite3_bind_text(res, 1, ticker.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 2, a.title.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 3, a.source.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 4, a.published_at.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 5, a.url.c_str(), -1, SQLITE_STATIC);
            sqlite3_bind_text(res, 6, a.snippet.c_str(), -1, SQLITE_STATIC);
            sqlite3_step(res);
            sqlite3_reset(res);
        }
        sqlite3_finalize(res);
    }
}
